File: benchmarks/compare/raptor/python/raptor/scan_results.py
"""ScanResults + ScanResultsFactory (mirrors src/raptor/ScanResults.ts).

Connections are either a Transfer or a (Trip, startIndex, endIndex) tuple.
All maps are insertion-ordered dicts (contract #1). getMarkedStops returns the
insertion order of the current round's arrivals.
"""
from __future__ import annotations

from typing import Dict, List, Tuple, Union

from .gtfs import MAX_SAFE_INTEGER, StopID, Time, Transfer, Trip

Connection = Tuple[Trip, int, int]
# A kConnections entry value is keyed by round number k -> Connection | Transfer.
ConnectionIndex = Dict[StopID, Dict[int, Union[Connection, Transfer]]]
Arrivals = Dict[StopID, Time]


class ScanResults:
    def __init__(
        self,
        best_arrivals: Arrivals,
        k_arrivals: Dict[int, Arrivals],
        k_connections: ConnectionIndex,
    ) -> None:
        self.k = 0
        self.bestArrivals = best_arrivals
        self.kArrivals = k_arrivals
        self.kConnections = k_connections

    def addRound(self) -> None:
        self.k += 1
        self.kArrivals[self.k] = {}

    def previousArrival(self, stop_pi: StopID) -> Time:
        # JS returns undefined when absent; callers test truthiness, so use None.
        return self.kArrivals[self.k - 1].get(stop_pi)

    def setTrip(self, trip: Trip, start_index: int, end_index: int, interchange: int) -> None:
        time = trip.stopTimes[end_index].arrivalTime + interchange
        stop_pi = trip.stopTimes[end_index].stop

        self.kArrivals[self.k][stop_pi] = time
        self.bestArrivals[stop_pi] = time
        self.kConnections[stop_pi][self.k] = (trip, start_index, end_index)

    def setTransfer(self, transfer: Transfer, time: Time) -> None:
        stop_pi = transfer.destination

        self.kArrivals[self.k][stop_pi] = time
        self.bestArrivals[stop_pi] = time
        self.kConnections[stop_pi][self.k] = transfer

    def bestArrival(self, stop_pi: StopID) -> Time:
        return self.bestArrivals[stop_pi]

    def getMarkedStops(self) -> List[StopID]:
        return list(self.kArrivals[self.k])

    def finalize(self) -> Tuple[ConnectionIndex, Arrivals]:
        return self.kConnections, self.bestArrivals


class ScanResultsFactory:
    def __init__(self, stops: List[StopID]) -> None:
        self.stops = stops

    def create(self, origins: Dict[StopID, Time]) -> ScanResults:
        best_arrivals = {stop: origins.get(stop) or MAX_SAFE_INTEGER for stop in self.stops}
        k_arrivals = [{stop: origins.get(stop) or MAX_SAFE_INTEGER for stop in self.stops}]
        # kArrivals is keyed by round number; round 0 holds the origins.
        k_arrivals_map = {0: k_arrivals[0]}
        k_connections: ConnectionIndex = {stop: {} for stop in self.stops}

        return ScanResults(best_arrivals, k_arrivals_map, k_connections)

File: benchmarks/compare/raptor/node/src/raptor/ScanResults.js
const INF = 4_000_000_000;

export class ScanResults {
  constructor(numStops, bestArrivals, kArrivals, kConnections, stopIndexOf, stopNames) {
    this.k = 0;
    this.numStops = numStops;
    this.bestArrivals = bestArrivals;   // Float64Array[numStops]
    this.kArrivals = kArrivals;         // Array of Float64Array[numStops]
    this.kConnections = kConnections;   // {stopName: {round: connection}} for reconstruction
    this.stopIndexOf = stopIndexOf;
    this.stopNames = stopNames;
    // Marked stops this round (int indices)
    this._markedThisRound = [];
  }

  addRound() {
    this.k++;
    this.kArrivals.push(new Float64Array(this.numStops).fill(INF));
    this._markedThisRound = [];
  }

  previousArrival(stopIdx) {
    const v = this.kArrivals[this.k - 1][stopIdx];
    return v >= INF ? -1 : v;
  }

  setTrip(trip, startIndex, endIndex, interchangeVal, stopIdx, rawArrival) {
    const time = rawArrival + interchangeVal;
    this.kArrivals[this.k][stopIdx] = time;
    this.bestArrivals[stopIdx] = time;
    this._markedThisRound.push(stopIdx);

    const stopName = this.stopNames[stopIdx];
    this.kConnections[stopName][this.k] = [trip, startIndex, endIndex];
  }

  setTransfer(transfer, time, destIdx) {
    this.kArrivals[this.k][destIdx] = time;
    this.bestArrivals[destIdx] = time;
    this._markedThisRound.push(destIdx);

    const stopName = this.stopNames[destIdx];
    this.kConnections[stopName][this.k] = transfer;
  }

  bestArrival(stopIdx) {
    const v = this.bestArrivals[stopIdx];
    return v >= INF ? INF : v;
  }

  getMarkedStops() {
    // Return the list of int stop indices improved this round, deduplicated.
    // (A stop can be marked multiple times in one round via different routes/transfers.)
    if (this._markedThisRound.length <= 1) return this._markedThisRound.slice();
    return [...new Set(this._markedThisRound)];
  }

  finalize() {
    // Return kConnections keyed by string stop ids (already is), plus bestArrivals
    // converted back to a string-keyed map for GroupStationDepartAfterQuery.getFoundStations.
    const bestArrivalsMap = {};
    const stopNames = this.stopNames;
    const ba = this.bestArrivals;
    for (let i = 0; i < this.numStops; i++) {
      if (ba[i] < INF) bestArrivalsMap[stopNames[i]] = ba[i];
    }
    return [this.kConnections, bestArrivalsMap];
  }
}
